bip-taproot.mediawiki

```diff
@@ -59,19 +59,19 @@ The following rules only apply when such an output is being spent. Any other out
 ** Let ''p = c[1:33]'' and let ''P = point(p)'' where ''point'' and ''[:]'' are defined as in bip-schnorr. Fail if this point is not on the curve.
-** Let ''l = c[0] & 0xfe'', the leaf version<ref>'''What is the purpose of the first byte of the control block?''' The first byte of the control block has three distinct functions:
+** Let ''l = (c[0] ^ 0xc0) >> 1'', the leaf version<ref>'''What is the purpose of the first byte of the control block?''' The first byte of the control block has three distinct functions:
 * The low bit is used to denote whether the point represented by public key ''q'' is negated before verifying the taproot tweak.<ref>'''Why is it necessary to reveal a bit to indicate if the point represented by the output public key is negated in a script path spend?''' The ''point'' function (defined in bip-schnorr) always constructs a point with a square Y coordinate, but because ''Q'' is constructed by adding the taproot tweak to the internal public key ''P'', it cannot easily be guaranteed that ''Q'' in fact has such a Y coordinate. Therefore, before verifying the taproot tweak the original point is restored by negating if necessary. We can not ignore the Y coordinate because it would prevent batch verification. Trying out multiple internal keys until there's such a ''Q'' is possible but undesirable and unnecessary since this information about the Y coordinate only consumes an unused bit.</ref>
-* By keeping the top two bits set to true, it can be guaranteed that scripts can be recognized without knowledge of the UTXO being spent, simplifying analysis. This is because such values cannot occur as first byte of the final stack element in either P2WPKH or P2WSH spends.
-* The remaining five bits are used for introducing new script versions that are not observable unless actually executed.
+* The top two bits are set to true, which guarantees that the witness stack for a Taproot spend can be distinguished from a P2WPKH or P2WSH spend. For P2WPH, the last witness stack item is a public key, whose first byte is either 0x02 or 0x03. For P2WSH, the last item is a script, and scripts must contain valid opcodes, but no value &ge; 0xc0 is a valid opcode. The ability to recognize Taproot spends without access to the output being spent may be useful for some types static analysis. Note that leaf versions above 31 lack this property.
```

Comment:
typo p2wph

Reply:
This seems a bit confusing; maybe rearrange to put the purpose first, and the details after?

> Some types of static analysis may benefit from the ability to analyse spends without access to the output being spent. We achieve that for leaf versions 0 to 31 via flipping the top two bits of the first byte of the control block and observing that this ensures they can be distinguished from P2WPKH and P2WSH spends -- as for P2WPH, the last witness stack item is a public key, whose first byte is either 0x02 or 0x03. For P2WSH, the last item is a script, and scripts must contain valid opcodes, but no value &ge; 0xc0 is a valid opcode.

Reply:
Fixed typo by switching to @ajtowns' text.

Reply:
Fixed the P2WPH typo by switching to @ajtowns' text.